content/posts/vmware-esx-memory-management.md

```diff
@@ -0,0 +1,46 @@
+
+### Data sharing
+
+When multiple guests share the same data (code pages, read-only data), ESX deduplicates the data by adding another layer of indirection from a "physical" (to the operating system) page to a machine (managed by ESX) page. This allows the hypervisor to deduplicate the pages when they are mapped in by hashing the contents of a page and if the hashes match, verifying the content actually matches before remapping the "physical" page. ESX also performs copy-on-write (CoW) so when a guest writes to a shared page, CoW occurs and preserves correctness. However, this becomes untenable at higher-scale due to $O(n^2)$ page search, so ESX picks random times to deduplicate pages. In testing, up to 60% of a guest's memory can be deduplicated, though this is the optimistic case - diverse workloads fared worse due to the time required to check for potential deduplication.
+
+### Resource sharing
+
+ESX uses a ticket/lease mechanism to distribute resources to guests - the more tickets a guest has, the more resources available to them. To prevent a guest from hogging/acquiring all resources and refusing to release them, ESX implements the concept of the "idle memory tax"; the tax comes at reclamation time - guests with more unused resources will be picked first to have resources taken away from them.
+
+## Benchmarks
+
+Many benchmarks were done - refer to the paper for methodology, equations, and results. The readers/author of this summary do not have the pre-requisite background to accurately interpret those results.
+
+## Future Work
+
+In the time since the paper was released, there does not appear to have been much advancement in this area - Linux/KVM [still uses](https://pmhahn.github.io/virtio-balloon/) memory ballooning. At a higher-level, containerization has taken over many of the old use cases of virtual machines and in that case, a regular operating system kernel can rely on traditional OS techniques to do resource management. Additionally, we hypothesize many of these techniques are "good enough" and no one has felt motivated to significantly improve on these techniques.
```

Comment:
Aren't containers mostly run on VMs offered by cloud providers (like GCE)?

Reply:
Good call-out, many of them are. It seems like the hyperscalers use smaller hypervisors in cases like GCE/GKE - see [gVisor](https://gvisor.dev/docs/) (Google) and [Firecracker](https://firecracker-microvm.github.io/) (Amazon). It looks like they minimize KVM usage/remove it altogether.